weblogic-batch/admin/jms/report-stuck-ef-submissions.sh

```diff
@@ -0,0 +1,147 @@
+source /apps/oracle/scripts/logging_functions
+
+exec > >(tee "${LOG_FILE}") 2>&1
+
+TMP_LIST_FILE=list-all-jms.output
+TMP_METADATA_FILE=extracted-metadata.tmp
+TMP_XML_FILE=xml.tmp
+
+# Extract a list of submission xml from all jms servers
+./list-all-jms.sh uk.gov.ch.chips.jms.EfilingRequestErrorQueue > ${TMP_LIST_FILE}
+
+# Tidy up the extracted list so that it just contains xml
+sed -i -n '/^<?xml/,/<\/form>/p' ${TMP_LIST_FILE}
+
+# Check if there is any xml in TMP_LIST_FILE - if not then we can exit as no stuck docs
+grep -q xml ${TMP_LIST_FILE}
+if [[ $? -gt 0 ]]
+then
+  echo "No stuck documents"
+  exit 0
+fi
+
+# Extract data from the list of stuck docs in TMP_LIST_FILE, one document at a time
+rm -f ${TMP_XML_FILE} ${TMP_METADATA_FILE}
+while read -r LINE; do
+  if [[ ${LINE} =~ ^\<\?xml ]]
+  then
+    # new xml found - process previously saved xml if present
+    if [[ -f ${TMP_XML_FILE} ]]
+    then
+      parseAndWriteMetaData ${TMP_XML_FILE} ${TMP_METADATA_FILE}
+    fi
+  fi
+  echo "${LINE}" >> ${TMP_XML_FILE}
+
+done <${TMP_LIST_FILE}
+# Process the last submission as we have reached the end of TMP_LIST_FILE
+parseAndWriteMetaData ${TMP_XML_FILE} ${TMP_METADATA_FILE}
+
+# Now we have our metadata, we need to generate a report
+
+TMP_REPORT_FILE=report.tmp
+rm -f  ${TMP_REPORT_FILE}
+
+COUNT=$(grep -c -v scan ${TMP_METADATA_FILE})
+if [[ ${COUNT} -gt 0 ]]
+then
+  { 
+  smallBanner "EF/CHS docs"
+
+  echo "  Cnt | Type"
+  grep -v scan ${TMP_METADATA_FILE} | sort | awk -F\| '{print $1}' | uniq -c
+
+  echo
+  IFS=$'\n'
+  for DOC_TYPE in $(grep -v scan ${TMP_METADATA_FILE} | sort | awk -F\| '{print $1}' | uniq)
+  do
+    echo "  ${DOC_TYPE}:"
+    grep -v scan ${TMP_METADATA_FILE} | grep -E "^${DOC_TYPE}" | sed 's/[^|]*|\(.*\)/  \1/' | sed 's/|/    /g'
+    echo
+  done
+  } >> ${TMP_REPORT_FILE}
+fi
+
+
+COUNT=$(grep -c scan ${TMP_METADATA_FILE})
+if [[ ${COUNT} -gt 0 ]]
+then
+  {
+  smallBanner "FES/Scanned docs"
+
+  echo "  Cnt | Type"
+  grep scan ${TMP_METADATA_FILE} | sort | awk -F\| '{print $1}' | uniq -c
+
+  echo
+  IFS=$'\n'
+  for DOC_TYPE in $(grep scan ${TMP_METADATA_FILE} | sort | awk -F\| '{print $1}' | uniq)
+  do
+    echo "  ${DOC_TYPE}:"
+    grep scan ${TMP_METADATA_FILE} | grep -E "^${DOC_TYPE}" | sed 's/[^|]*|\(.*\)/  \1/' | sed 's/|/    /g'
+    echo
+  done
+  } >> ${TMP_REPORT_FILE}
+fi
+
+# Now we need to email out the report
+
+LOCATION_STRING="Sent from: ${ENVIRONMENT_LABEL} - ${T3_HOST_FQDN} - EC2 instance ID ${EC2_INSTANCE_ID} (${APP_INSTANCE_NAME})"
+DATE=$(date)
+SUBJECT="Following CLOUD EF or FES/Scanned docs currently stuck ${DATE}"
+HEADING="Snapshot of docs in EfilingRequestErrorQueue at ${DATE}. CSI will fix these and push these to QH queues ASAP."
+
+TMP_EMAIL_FILE=email.tmp
+echo -e "To:${EMAIL_ADDRESS_CSI}\nFrom:${EMAIL_ADDRESS_CSI}\nSubject:${SUBJECT}\n\n${LOCATION_STRING}\n\n${HEADING}" > ${TMP_EMAIL_FILE}
+cat ${TMP_REPORT_FILE} >> ${TMP_EMAIL_FILE}
+cat ${TMP_EMAIL_FILE} | msmtp -t
```

Comment:
Should we try to use the alert_functions script for all emails for maintainability, i.e. only having to edit it in one place if we change the mechanism we use in future?
(letter-producer/bad_letters.sh has an example of sending file contents in an email report)